[fix] Limit default max zoom level on mapOptions #188

dee077 opened the pull request:
Prevent map from exceeding supported zoom levels of tile providers and added cursor not allowed on hitting min or max zoom level

Fixes #188

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [ ] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

## Reference to Existing Issue

Closes #188.

## Description of Changes

- Added minZoom:3 and maxZoom:18 to mapOptions
- Reduced maxZoom to 18 in mapTileConfig
- Added cursor not allowed on hitting max or min zoom level
- Updated the same zoom levels in tests

## Demo


https://github.com/user-attachments/assets/ba7b2e6b-c36e-4083-9620-fbb112e5827a

nemesifier reviewed (commented):
@dee077 thanks for adding the recording! It looks very promising!
I will test and look closer into this asap. :+1: :+1:

niteshsinha17 commented on src/js/netjsongraph.render.js:
```diff
@@ -456,26 +456,33 @@ class NetJSONGraphRender {
+      const zoomOutBtn = document.querySelector(".leaflet-control-zoom-out");
+
+      if (currentZoom >= maxZoom) {
+        zoomInBtn.style.cursor = "not-allowed";
```
instead of cursor not allowed, let's reduce its opacity and set pointer event to none.

niteshsinha17 replied:
We can have classname for this.

dee077 replied:
Sure

dee077 commented:
### Updates
- Take zoom values from `self.config`
- Toggle zoom button state via classes instead of directly changing styles
- Set floorplan default zoom to 4 (to avoid decimal value conflicts in <= comparison)

nemesifier commented on src/js/netjsongraph.render.js:
```diff
@@ -456,26 +456,36 @@ class NetJSONGraphRender {
+      const currentZoom = self.leaflet.getZoom();
+      const {maxZoom} = self.config.mapOptions;
+      const {minZoom} = self.config.mapOptions;
+      console.log(minZoom, maxZoom);
```
```suggestion
```

nemesifier commented on src/js/netjsongraph.config.js:
```diff
@@ -238,7 +240,7 @@ const NetJSONGraphDefaultConfig = {
       options: {
         minZoom: 3,
-        maxZoom: 32,
+        maxZoom: 18,
```
Why do we need minZoom and maxZoom to be defined both here and above in `mapOptions`?

Can we avoid this redundancy? Maybe the `mapTileConfig` logic can take these settings from `mapOptions`?
Or would it be possible in each `mapTileConfig` to override what is in `mapOptions`? If so, a few code comments would help understanding this more quickly.

I see the minZoom and maxZoom properties of `mapTileConfig` are already documented in the README, please review the text there to ensure it's up to date.

nemesifier reviewed (changes requested):
@dee077 looks almost ready but please see my comments below.

Can you add 1 browser test for testing maxZoom and minZoom?

dee077 commented:
### Updates:
- Removed min max zoom from `mapTileConfig`.
- Added selenium test to check the max zoom. The the tile error coming on console is coming late after in between 5s to 10s. That's why added a sleep time of 15s to catch these errors. 
- Sometimes the test runs slower in ci, should I increase the sleep time to more?

nemesifier reviewed (commented):
@dee077 @niteshsinha17 this looks good to me now, the only issue is the coverage going down, the browser tests now are excluded from the coverage calculation, do you know if there's a way to include them?

niteshsinha17 commented on .github/workflows/ci.yml:
```diff
@@ -85,11 +85,15 @@ jobs:
       - name: Unit tests
-        run: yarn coverage --testPathIgnorePatterns=test/netjsongraph.browser.test.js
-  
+        run: npx nyc --silent yarn coverage --testPathIgnorePatterns=test/netjsongraph.browser.test.js
```
> @dee077 @niteshsinha17 this looks good to me now, the only issue is the coverage going down, the browser tests now are excluded from the coverage calculation, do you know if there's a way to include them?

Looks good to me too.

@dee077 @nemesifier I think `netjsongraph.browser.test.js` is being ignored while running tests and generating coverage. May be need to change this line.

niteshsinha17 reviewed (commented):
Check my comment below.

nemesifier reviewed (commented):
BTW I think this is part of the GSoC project of @cestercian.

I would let @cestercian step in on this. Yash, please review and test.
Can you find a solution for the test coverage?

cestercian commented:
### Updates:

- Removed enabled disabled class addition using js as there's already a class leaflet disabled and leaflet enabled enforced on enabled disabled state 

- Updated test cases for the same

nemesifier reviewed (commented):
Help me understand what's going on, the test coverage now doesn't suffer due to the removal of the custom code?

cestercian commented:
I removed the custom JS that manually toggled zoom button states and instead used Leaflet’s native `.leaflet-disabled` class (which I discovered while going through the Leaflet docs).

Since there's no custom logic anymore, coverage stays the same because we're just testing config now, not redundant JS.

dee077 commented:
I refactored the code to assign values conditionally instead of using separate if-else blocks in render.js. This reduced the number of uncovered lines in the file, while keeping the tests unchanged. As a result, we observed a slight increase in the coverage percentage.

Previously, when we added custom classes for enabled and disabled states, it introduced additional uncovered lines, which led to a drop in coverage.

But still the selenium tests are not included in the coverage.

nemesifier reviewed (commented):
I found an issue while testing: clicking on the disabled button seems to move the map, as if the button became transparent with respect to its behavior.

Is that intended? Can it be prevented? Ideally nothing should happen.

[max-zoom.webm](https://github.com/user-attachments/assets/b2171631-3b80-4526-997a-6761e1a87c11)

cestercian commented:
https://github.com/user-attachments/assets/108fac37-d51d-4dd2-a721-70bcae330658

`pointer-events: none` makes the element transparent to pointer interactions — including allowing clicks to pass through to elements underneath. At the time, we actually thought that behavior was beneficial for our use case.

That said, it turned out to cause unintended side effects (like interacting with the map beneath a disabled button), so I’ve removed the property to keep things working as expected. 🙂

nemesifier reviewed (approved):
Great work @dee077 @cestercian! :+1: 

I am assuming the `mapTileConfig` line in the README is still valid also after these changes, right? If not, please update it.
